lib/utils/deeply-non-duplicated-addon.js

```diff
@@ -0,0 +1,44 @@
+'use strict';
+
+/**
+ * Deduplicate one addon's children addons recursively from hostAddons.
+ *
+ * @private
+ * @param {Object} hostAddons
+ * @param {EmberAddon} dedupedAddon
+ * @param {String} treeName
+ */
+module.exports = function deeplyNonDuplicatedAddon(hostAddons, dedupedAddon, treeName) {
+  if (dedupedAddon.addons.length === 0) {
+    return;
+  }
+
+  dedupedAddon._orginalAddons = dedupedAddon.addons;
```

Comment:
Although this works in common situation, I believe it poses some broader risks. For example, if the patched `addon` instance is referred to in async or callback code, it will appear to have divergent `addons` arrays.

Although this scenario is rare, when it occurs it will be extremely surprising:

Contrived example to illustrate:

```js
treeForSomething(tree) {
  let addonsWhenHookCalled = this.addons;
  return new BroccoliPlugin(tree, {
    someCallbackInvokedLater: () => {
      let addons = this.addons; // imagine we need the list of addons here for some reason

      addons !== addonsWhenHookCalled; // this becomes surprising and incoherent
     }
  });
}
```

---

I've thought about how to fix this, I believe it can be done by utilizing a proxy, which wrap's an add-on, and traps when `addon.addons` is accessed, providing only the addons that should be visible. (And doing so recursively).

Does taking this approach sound reasonable?

Reply:
Could any `treeForSomething` in `engine-addon.js` or ember-lic addon have async behaviour? It's seems everything is sync like `return addon[methodName].apply(addon, invokeArguments);` or it should return a Promise.
Is there any real world example rather than a contrived one?

Reply:
@2hu12 contrived, and for us to discuss. If we don't feel it is worth it, we don't need to doit.